Fix highlighting warning and error issues

jorgekorgut opened the pull request:
Fix Issue #104413 : Add a limit of 20 lines of warning highlight otherwise paint the background of the first line only.
![highlight-limit-background](https://github.com/user-attachments/assets/cc8b536f-707c-4668-abb1-828460fa8b37)
Fix Issue #106278 : Extract the logic from clearing background lines from ```_update_errors()``` to ```_clear_background_lines()```
Fix Issue #83979 : It was fixed on 4.5.dev instead of 4.2.beta.3
![highlight-incoherence-unfold-all](https://github.com/user-attachments/assets/66822f75-8145-4165-9724-7fc6271281d5)

Feature: Highlight the folded line using inline events. The background is set in the following priority order: Error, Warning, then Highlight background. Also added new calls to ```_validate_scripts()``` to keep the editor visuals up to date.
![highlight-fold-feature](https://github.com/user-attachments/assets/94afa37c-788a-4302-92c2-81822f531b2a)


*Maintainer edit*:
- Fixes #104413
- Fixes #83979

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1806,6 +1806,18 @@ void CodeEdit::toggle_foldable_lines_at_carets() {
 
+int CodeEdit::get_folded_line_header(int p_line) const {
+	ERR_FAIL_INDEX_V(p_line, get_line_count(), false);
+	// Search for the first non hidden line
```
```suggestion
	// Search for the first non hidden line.
```

jorgekorgut commented:
Thanks for reviewing, I will post a new pull request with the modifications soon.

jorgekorgut commented:
I just finished implementing your sugestions.  
Note that I refactor: ```_update_background_text()``` to extracting the background painting logic from the ```_validate_script()``` process.

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -605,6 +605,48 @@ void ScriptTextEditor::_update_color_constructor_options() {
 	}
 }
 
+void ScriptTextEditor::_update_background_text() {
```
```suggestion
void ScriptTextEditor::_update_background_colors() {
```

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -605,6 +605,48 @@ void ScriptTextEditor::_update_color_constructor_options() {
+
+	// Set the warning background.
+	if (warning_line_color.a != 0.0) {
+		for (const ScriptLanguage::Warning &W : warnings) {
```
```suggestion
		for (const ScriptLanguage::Warning &warning : warnings) {
```

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -605,6 +605,48 @@ void ScriptTextEditor::_update_color_constructor_options() {
+	}
+
+	// Set the error background.
+	for (const ScriptLanguage::ScriptError &E : errors) {
```
```suggestion
	if (marked_line_color.a != 0.0) {
		for (const ScriptLanguage::ScriptError &error : errors) {
```
This should check for alpha, and `error` is more clear.

kitbdev commented:
I'm not sure why the warnings are like that, it seems like that is how the GDScript parser handles multiline nodes.

> And in the code_edit.cpp we have 4 references of the correction from false to 0 that you spotted. Should it also be corrected ?

Those other ones look correct, since the return types of those functions are bool.

> Fix Issue https://github.com/godotengine/godot/issues/83979 : It was fixed on 4.5.dev instead of 4.2.beta.3

I'm not sure what you mean, the issue wasn't fixed in 4.2.beta.3, it was reported then.

Technically, this doesn't close https://github.com/godotengine/godot/issues/83979, since a CodeEdit by itself outside the Editor can still call `unfold_all_lines` and the highlight doesn't clear.
It does work here now in the ScriptEditor.

kitbdev reviewed (commented):
Working as expected.

jorgekorgut commented:
> I'm not sure what you mean, the issue wasn't fixed in 4.2.beta.3, it was reported then.

I am not sure how to handle an issue that was spotted in a previous version but not corrected. Should I make a commit in a branch that the base is close from the tag(4.2), or should I just do what I have done in the most recent master branch ?

kitbdev commented:
> I am not sure how to handle an issue that was spotted in a previous version but not corrected. Should I make a commit in a branch that the base is close from the tag(4.2), or should I just do what I have done in the most recent master branch ?

What you did is correct, targetting the master branch. 4.2 doesn't really receive updates anymore, unless its a critical fix.

> Do you think I should keep digging to solve the issue https://github.com/godotengine/godot/issues/83979 in this pull request or should I left it out and make another pull request later for it ?

It can be in another PR later.

jorgekorgut commented:
I have another proposal : What do you think if we emit a signal inside ```CodeEdit``` when folding/unfolding and in the ```ScriptEditor``` we call ``` _update_background_color()``` on its signal ?

kitbdev commented:
> I have another proposal : What do you think if we emit a signal inside `CodeEdit` when folding/unfolding and in the `ScriptEditor` we call ` _update_background_color()` on its signal ? It would remove the multiple explicit independent calls of `_update_background_color()` from the script editor.

Actually, yes there should be some kind of fold updated signal, since there are other ways a line can be folded/unfolded that are currently not handled (like the unfold button at the end of a folded line).

jorgekorgut commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1731,11 +1753,9 @@ void ScriptTextEditor::_edit_option(int p_op) {
 		} break;
 		case EDIT_UNFOLD_ALL_LINES: {
 			tx->unfold_all_lines();
-			tx->queue_redraw();
```
this ```tx->queue_redraw()``` were called once inside ```tx->fold_all_lines()``` and another one here.

jorgekorgut commented on scene/gui/code_edit.cpp:
```diff
@@ -1672,7 +1672,7 @@ bool CodeEdit::can_fold_line(int p_line) const {
 }
 
-void CodeEdit::fold_line(int p_line) {
+void CodeEdit::_fold_line(int p_line) {
```
I change for a private call for ```_fold_line()``` that does not calls ```queue_redraw()``` and emit the signal ```fold_line_updated```.

jorgekorgut commented on scene/gui/code_edit.cpp:
```diff
@@ -1697,7 +1697,6 @@ void CodeEdit::fold_line(int p_line) {
 				}
 			}
 		}
-		set_line_background_color(p_line, theme_cache.folded_code_region_color);
```
Now that the background color is updated in the ```script_text_editor```, we dont need this call here. Or do we ?

kitbdev replied:
CodeEdit still needs to handle highlighting itself, for when it is not part of the ScriptEditor and just by itself.

jorgekorgut commented on scene/gui/code_edit.cpp:
```diff
@@ -1762,22 +1761,38 @@ void CodeEdit::unfold_line(int p_line) {
+	emit_signal(SNAME("fold_line_updated"));
 }
 
 void CodeEdit::unfold_all_lines() {
```
standardize the ```unfold_all_lines()``` to call ```_unfold_line()``` for every line.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -7580,6 +7580,7 @@ void TextEdit::_bind_methods() {
 	ADD_SIGNAL(MethodInfo("text_set"));
 	ADD_SIGNAL(MethodInfo("text_changed"));
 	ADD_SIGNAL(MethodInfo("lines_edited_from", PropertyInfo(Variant::INT, "from_line"), PropertyInfo(Variant::INT, "to_line")));
+	ADD_SIGNAL(MethodInfo("fold_line_updated"));
```
This should be in CodeEdit, not TextEdit.

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1762,22 +1761,38 @@ void CodeEdit::unfold_line(int p_line) {
+	for (int i = 0; i < get_line_count(); i++) {
+		_unfold_line(i);
+	}
+	// Missing _update_scrollbars() call to be exactly the same.
+	queue_accessibility_update();
+	queue_redraw();
```
```suggestion
```
You shouldn't have to queue the accessibility update or queue redraw for any of the line folding or unfolding, since `_set_line_as_hidden` handles it. And don't worry that it gets called multiple times, since its queued.
The `_update_scrollbars`, you probably don't have to worry about, as long as it seems like the scrollbars get updated when folding/unfolding.

kitbdev reviewed (commented):
Functionality looks good.
Looks like you need to run [doctool](https://docs.godotengine.org/en/latest/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine) and add a description for the new signal. Something like `Emitted when lines are folded or unfolded.`

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1766,18 +1768,47 @@ void CodeEdit::unfold_line(int p_line) {
+
 	for (int i = 0; i < get_line_count(); i++) {
-		fold_line(i);
+		any_line_folded != _fold_line(i);
```
```suggestion
		any_line_folded |= _fold_line(i);
```

jorgekorgut replied:
sorry 😐

jorgekorgut commented on scene/gui/code_edit.cpp:
```diff
@@ -2880,6 +2923,9 @@ void CodeEdit::_bind_methods() {
 	ADD_SIGNAL(MethodInfo("breakpoint_toggled", PropertyInfo(Variant::INT, "line")));
 
+	/* Fold Lines */
+	ADD_SIGNAL(MethodInfo("fold_line_updated"));
```
isn't the same case for some signals that are already in the codebase ?
```symbol_hovered```  
```code_completion_requested```  

From the conversations I had with @kitbdev, I understood that ```script_text_editor``` is not the only class that should use the ```code_edit``` functionalities.

Mickeon replied:
The "danger" is having one more piece of public API to maintain.
If this signal is deemed obsolete it cannot be easily removed. If it were to require more parameters, it cannot be changed because that would break compatibility, and there's no way to handle it without adding a whole other signal.

Once again, unfortunately, there's no way to prevent signals from being exposed and documented.

Related discussion: https://github.com/godotengine/godot/pull/69012#discussion_r2051736104 and https://github.com/godotengine/godot/pull/102380#discussion_r2060648952

jorgekorgut replied:
Thanks for taking the time to explain !

kitbdev replied:
I think this signal is useful enough to be public, since there is currently no good way to respond to changes in folding/unfolding since it can happen in multiple places. 
Parameters could be added to it to specify a line range, but I don't know how necessary they would be.

Also it looks like some editor only signals have been added recently, not sure if it is valid: https://github.com/godotengine/godot/pull/101427

https://github.com/godotengine/godot/blob/a2aefab4c78273c8a2006daf28b700d427e1feeb/scene/2d/camera_2d.cpp#L1064-L1066

jorgekorgut replied:
What do you think then @Mickeon, @kitbdev, to make this signal private (with the starting underline) and if someones claims the need to this function public later, it is easy to change. 
If it becomes obsolete and used outside of its scopes, this signal was discourage by how it was implemented and then it is not the engines fault the lack of compatibility.

jorgekorgut committed:
Fix highlighting warning and error issues related to Godot editor.

Fix Issue #104413 : Add a limit of 20 lines of warning highlight otherwise paint the background of the first line only.
Fix Issue #106278 : Extract the logic from clearing and setting background lines from _update_errors() and _update_warnings() to _update_background_text().
Fix Issue #83979 : Added signal for folding/unfolding lines (fold_lines_updated).
Feature: Highlight the folded line using signals connected to _update_background_text() to keep the editor visuals up to date. The background is set in the following priority order: Error, Warning, then Highlight background.

kitbdev reviewed (approved):
Looks good, fixes the highlighting issues.
